[WebXR] Add bindings and stubs of Hit Test Module

fujii opened the pull request:
#### 7ab677a27cf3c8b64ffc7cac7fdefd5a9c535635
<pre>
[WebXR] Add bindings and stubs of Hit Test Module
<a href="https://bugs.webkit.org/show_bug.cgi?id=300865">https://bugs.webkit.org/show_bug.cgi?id=300865</a>

Reviewed by Sergio Villar Senin.

Added IDLs and stubs for WebXR Hit Test Module
&lt;<a href="https://immersive-web.github.io/hit-test/">https://immersive-web.github.io/hit-test/</a>&gt;. Added a new macro
ENABLE_WEBXR_HIT_TEST which is ON only for GTK and WPE developer mode. Added a
new preference WebXRHitTestModuleEnabled which is off by default.

Tests: imported/w3c/web-platform-tests/webxr/hit-test/idlharness.https.html

* LayoutTests/imported/w3c/web-platform-tests/webxr/hit-test/idlharness.https-expected.txt: Added.
* LayoutTests/platform/glib/TestExpectations:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WTF/wtf/PlatformEnable.h:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Modules/webxr/WebXRFrame+HitTest.idl: Added.
* Source/WebCore/Modules/webxr/WebXRFrame.cpp:
(WebCore::WebXRFrame::getHitTestResults):
(WebCore::WebXRFrame::getHitTestResultsForTransientInput):
* Source/WebCore/Modules/webxr/WebXRFrame.h:
* Source/WebCore/Modules/webxr/WebXRHitTestResult.cpp: Added.
(WebCore::WebXRHitTestResult::getPose):
* Source/WebCore/Modules/webxr/WebXRHitTestResult.h: Added.
* Source/WebCore/Modules/webxr/WebXRHitTestResult.idl: Added.
* Source/WebCore/Modules/webxr/WebXRHitTestSource.cpp: Added.
(WebCore::WebXRHitTestSource::create):
(WebCore::WebXRHitTestSource::cancel):
* Source/WebCore/Modules/webxr/WebXRHitTestSource.h: Added.
* Source/WebCore/Modules/webxr/WebXRHitTestSource.idl: Added.
* Source/WebCore/Modules/webxr/WebXRRay.cpp: Added.
(WebCore::WebXRRay::create):
(WebCore::WebXRRay::WebXRRay):
(WebCore::WebXRRay::origin):
(WebCore::WebXRRay::direction):
(WebCore::WebXRRay::matrix):
* Source/WebCore/Modules/webxr/WebXRRay.h: Added.
* Source/WebCore/Modules/webxr/WebXRRay.idl: Added.
* Source/WebCore/Modules/webxr/WebXRSession+HitTest.idl: Added.
* Source/WebCore/Modules/webxr/WebXRSession.cpp:
(WebCore::WebXRSession::requestHitTestSource):
(WebCore::WebXRSession::requestHitTestSourceForTransientInput):
* Source/WebCore/Modules/webxr/WebXRSession.h:
* Source/WebCore/Modules/webxr/WebXRSystem.cpp:
(WebCore::WebXRSystem::isFeaturePermitted const):
* Source/WebCore/Modules/webxr/WebXRTransientInputHitTestResult.cpp: Added.
(WebCore::WebXRTransientInputHitTestResult::create):
(WebCore::WebXRTransientInputHitTestResult::WebXRTransientInputHitTestResult):
(WebCore::WebXRTransientInputHitTestResult::inputSource const):
(WebCore::WebXRTransientInputHitTestResult::results const):
* Source/WebCore/Modules/webxr/WebXRTransientInputHitTestResult.h: Added.
* Source/WebCore/Modules/webxr/WebXRTransientInputHitTestResult.idl: Added.
* Source/WebCore/Modules/webxr/WebXRTransientInputHitTestSource.cpp: Added.
(WebCore::WebXRTransientInputHitTestSource::create):
(WebCore::WebXRTransientInputHitTestSource::cancel):
* Source/WebCore/Modules/webxr/WebXRTransientInputHitTestSource.h: Added.
* Source/WebCore/Modules/webxr/WebXRTransientInputHitTestSource.idl: Added.
* Source/WebCore/Modules/webxr/XRHitTestOptionsInit.h: Added.
* Source/WebCore/Modules/webxr/XRHitTestOptionsInit.idl: Added.
* Source/WebCore/Modules/webxr/XRHitTestTrackableType.h: Added.
* Source/WebCore/Modules/webxr/XRHitTestTrackableType.idl: Added.
* Source/WebCore/Modules/webxr/XRRayDirectionInit.h: Added.
* Source/WebCore/Modules/webxr/XRRayDirectionInit.idl: Added.
* Source/WebCore/Modules/webxr/XRTransientInputHitTestOptionsInit.h: Added.
* Source/WebCore/Modules/webxr/XRTransientInputHitTestOptionsInit.idl: Added.
* Source/WebCore/Modules/webxr/XRWebGLSubImage.h:
* Source/WebCore/Sources.txt:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/platform/xr/PlatformXR.h:
(PlatformXR::parseSessionFeatureDescriptor):
(PlatformXR::sessionFeatureDescriptor):
* Source/WebKit/Shared/XR/PlatformXR.serialization.in:
* Source/WebKit/UIProcess/API/glib/WebKitXRPermissionRequest.cpp:
(toFeatureList):
(toWebKitXRSessionFeatures):
* Source/WebKit/UIProcess/API/glib/WebKitXRPermissionRequest.h.in:
* Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.cpp:
(WebKit::OpenXRCoordinator::getPrimaryDeviceInfo):
* Source/cmake/OptionsGTK.cmake:
* Source/cmake/OptionsWPE.cmake:
* Source/cmake/WebKitFeatures.cmake:
* Tools/MiniBrowser/gtk/BrowserTab.c:
(WebKitXRSessionFeaturesToString):

Canonical link: <a href="https://commits.webkit.org/301799@main">https://commits.webkit.org/301799@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5a2da8f1655b8190cb1650b8e242edcc6d80e122

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/126764 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/46406 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/37375 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/133717 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/78392 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/128635 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/47034 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/54943 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/96471 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/64503 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/129712 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/37635 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/113384 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/76990 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/36516 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/31568 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/77110 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/118810 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/107456 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/31868 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/136295 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/125228 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/53441 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/41125 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/104978 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/53935 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/109745 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/104681 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/50176 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/28518 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/50895 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/19877 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/53369 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/59168 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/158269 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/52628 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/39593 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/55963 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/54375 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

svillar commented on Source/WebCore/Modules/webxr/WebXRRay.cpp:
```diff
@@ -0,0 +1,78 @@
+    auto exceptionOrTransform = WebXRRigidTransform::create(origin, DOMPointInit { direction.x, direction.y, direction.z, direction.w });
+    if (exceptionOrTransform.hasException())
+        return exceptionOrTransform.releaseException();
+    return adoptRef(*new WebXRRay(exceptionOrTransform.returnValue()));
```
releaseReturnValue()

fujii replied:
Fixed.

svillar commented on Source/WebCore/Modules/webxr/WebXRSystem.cpp:
```diff
@@ -275,6 +275,9 @@ bool WebXRSystem::isFeaturePermitted(PlatformXR::SessionFeature feature) const
     case PlatformXR::SessionFeature::WebGPU:
+#if ENABLE(WEBXR_HIT_TEST)
+    case PlatformXR::SessionFeature::HitTest:
+#endif
```
Hit testing requires xr-spatial-tracking so we need to check if the feature is enabled in the permissions policy

fujii replied:
Fixed.

svillar reviewed (changes requested):
Looking great so far. Just have 2 comments

svillar commented on Source/WebCore/Modules/webxr/WebXRSession.cpp:
```diff
@@ -32,14 +32,18 @@
 #include "JSDOMPromiseDeferred.h"
+#include "JSWebXRHitTestSource.h"
 #include "JSWebXRReferenceSpace.h"
+#include "JSWebXRTransientInputHitTestSource.h"
```
These includes are likely causing the errors in the VisionOS builds BTW

fujii replied:
Fixed.

svillar reviewed (approved):
Great! Thanks

fujii committed:
[WebXR] Add bindings and stubs of Hit Test Module
https://bugs.webkit.org/show_bug.cgi?id=300865

Reviewed by Sergio Villar Senin.

Added IDLs and stubs for WebXR Hit Test Module
<https://immersive-web.github.io/hit-test/>. Added a new macro
ENABLE_WEBXR_HIT_TEST which is ON only for GTK and WPE developer mode. Added a
new preference WebXRHitTestModuleEnabled which is off by default.

Tests: imported/w3c/web-platform-tests/webxr/hit-test/idlharness.https.html

* LayoutTests/imported/w3c/web-platform-tests/webxr/hit-test/idlharness.https-expected.txt: Added.
* LayoutTests/platform/glib/TestExpectations:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WTF/wtf/PlatformEnable.h:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Modules/webxr/WebXRFrame+HitTest.idl: Added.
* Source/WebCore/Modules/webxr/WebXRFrame.cpp:
(WebCore::WebXRFrame::getHitTestResults):
(WebCore::WebXRFrame::getHitTestResultsForTransientInput):
* Source/WebCore/Modules/webxr/WebXRFrame.h:
* Source/WebCore/Modules/webxr/WebXRHitTestResult.cpp: Added.
(WebCore::WebXRHitTestResult::getPose):
* Source/WebCore/Modules/webxr/WebXRHitTestResult.h: Added.
* Source/WebCore/Modules/webxr/WebXRHitTestResult.idl: Added.
* Source/WebCore/Modules/webxr/WebXRHitTestSource.cpp: Added.
(WebCore::WebXRHitTestSource::create):
(WebCore::WebXRHitTestSource::cancel):
* Source/WebCore/Modules/webxr/WebXRHitTestSource.h: Added.
* Source/WebCore/Modules/webxr/WebXRHitTestSource.idl: Added.
* Source/WebCore/Modules/webxr/WebXRRay.cpp: Added.
(WebCore::WebXRRay::create):
(WebCore::WebXRRay::WebXRRay):
(WebCore::WebXRRay::origin):
(WebCore::WebXRRay::direction):
(WebCore::WebXRRay::matrix):
* Source/WebCore/Modules/webxr/WebXRRay.h: Added.
* Source/WebCore/Modules/webxr/WebXRRay.idl: Added.
* Source/WebCore/Modules/webxr/WebXRSession+HitTest.idl: Added.
* Source/WebCore/Modules/webxr/WebXRSession.cpp:
(WebCore::WebXRSession::requestHitTestSource):
(WebCore::WebXRSession::requestHitTestSourceForTransientInput):
* Source/WebCore/Modules/webxr/WebXRSession.h:
* Source/WebCore/Modules/webxr/WebXRSystem.cpp:
(WebCore::WebXRSystem::isFeaturePermitted const):
* Source/WebCore/Modules/webxr/WebXRTransientInputHitTestResult.cpp: Added.
(WebCore::WebXRTransientInputHitTestResult::create):
(WebCore::WebXRTransientInputHitTestResult::WebXRTransientInputHitTestResult):
(WebCore::WebXRTransientInputHitTestResult::inputSource const):
(WebCore::WebXRTransientInputHitTestResult::results const):
* Source/WebCore/Modules/webxr/WebXRTransientInputHitTestResult.h: Added.
* Source/WebCore/Modules/webxr/WebXRTransientInputHitTestResult.idl: Added.
* Source/WebCore/Modules/webxr/WebXRTransientInputHitTestSource.cpp: Added.
(WebCore::WebXRTransientInputHitTestSource::create):
(WebCore::WebXRTransientInputHitTestSource::cancel):
* Source/WebCore/Modules/webxr/WebXRTransientInputHitTestSource.h: Added.
* Source/WebCore/Modules/webxr/WebXRTransientInputHitTestSource.idl: Added.
* Source/WebCore/Modules/webxr/XRHitTestOptionsInit.h: Added.
* Source/WebCore/Modules/webxr/XRHitTestOptionsInit.idl: Added.
* Source/WebCore/Modules/webxr/XRHitTestTrackableType.h: Added.
* Source/WebCore/Modules/webxr/XRHitTestTrackableType.idl: Added.
* Source/WebCore/Modules/webxr/XRRayDirectionInit.h: Added.
* Source/WebCore/Modules/webxr/XRRayDirectionInit.idl: Added.
* Source/WebCore/Modules/webxr/XRTransientInputHitTestOptionsInit.h: Added.
* Source/WebCore/Modules/webxr/XRTransientInputHitTestOptionsInit.idl: Added.
* Source/WebCore/Modules/webxr/XRWebGLSubImage.h:
* Source/WebCore/Sources.txt:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/platform/xr/PlatformXR.h:
(PlatformXR::parseSessionFeatureDescriptor):
(PlatformXR::sessionFeatureDescriptor):
* Source/WebKit/Shared/XR/PlatformXR.serialization.in:
* Source/WebKit/UIProcess/API/glib/WebKitXRPermissionRequest.cpp:
(toFeatureList):
(toWebKitXRSessionFeatures):
* Source/WebKit/UIProcess/API/glib/WebKitXRPermissionRequest.h.in:
* Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.cpp:
(WebKit::OpenXRCoordinator::getPrimaryDeviceInfo):
* Source/cmake/OptionsGTK.cmake:
* Source/cmake/OptionsWPE.cmake:
* Source/cmake/WebKitFeatures.cmake:
* Tools/MiniBrowser/gtk/BrowserTab.c:
(WebKitXRSessionFeaturesToString):

Canonical link: https://commits.webkit.org/301799@main
